pulp_2to3_migration/app/models/base.py

```diff
@@ -11,24 +17,113 @@ class MigrationPlan(Model):
     """
     plan = JSONField()
+    _real_plan = None
+
+    @property
+    def plan_view(self):
+        """
+        Window to view the validated migration plan data through.
+        """
+        # Needs a MongoDB connection initialized beforehand to look for importers,
+        # verify existance of repositories and distributors, etc.
+        if not self._real_plan:
+            self._real_plan = _ValidatedMigrationPlan(self.plan)
+
+        return self._real_plan
+
+    def get_plugins(self):
+        """
+        Return a list of pulp2 plugins to migrate.
+        """
+        return self.plan_view.plugins_to_migrate
 
     def get_repositories(self):
         """
         Return a list of pulp2 repositories to migrate or empty list if all should be migrated.
         """
-        # TODO: get pulp2 repositories from the self.plan
-        return []
+        return self.plan_view.repositories_to_migrate
 
     def get_importers(self):
         """
         Return a list of pulp2 importers to migrate or empty list if all should be migrated.
         """
-        # TODO: get pulp2 importers from the self.plan
-        return []
+        return self.plan_view.importers_to_migrate
 
     def get_distributors(self):
         """
         Return a list of pulp2 distributors to migrate or empty list if all should be migrated.
         """
-        # TODO: get pulp2 distributors from the self.plan
-        return []
+        return self.plan_view.distributors_to_migrate
+
+
+class SchemaValidationFailure(Exception):
+    """
+    Exception to be thrown when validating the MigrationPlan.
+
+    e.g. Repository specified does not exist.
+    """
+    pass
+
+
+class _ValidatedMigrationPlan:
+    def __init__(self, migration_plan):
+        self.migration_plan = migration_plan
+
+        self.plugins_to_migrate = []
+        self.importers_to_migrate = []
+        self.distributors_to_migrate = []
+        self.repositories_to_migrate = []
+
+        self.validate_and_populate()
+
+    def validate_and_populate(self):
+        for plugin_data in self.migration_plan['plugins']:
+            self.plugins_to_migrate.append(plugin_data['type'])
+            if plugin_data.get('repositories'):
+                self._parse_repository_data(plugin_data.get('repositories'))
+            # protection
+
+    def _parse_repository_data(self, repository_data):
+        for repository in repository_data:
+            try:
+                repo = Repository.objects.get(repo_id=repository['pulp2_repository_id'])
+            except Repository.DoesNotExist:
+                raise SchemaValidationFailure(
+                    "Repository id '{}' specified does not exist in Pulp2".format(
+                        repository['pulp2_repository_id']
+                    ))
+
+            try:
+                importer = Importer.objects.get(repo_id=repo.repo_id)
+                self.importers_to_migrate.append(importer.id)
+            except Importer.DoesNotExist:
```

Comment:
i believe we should raise an error here as well, @dralley ?

Reply:
Since these importers are found only by discovery via the listed repositories, I didn't think that was necessary.  Unlike the Repository and Distributor IDs they are not actually specified by the user in the sync plan.

Reply:
in case this we are using importer of provided pulp2_repository_id, which will transferred into remote in pulp3 terms.
it is true that it is quite unlikely to find a repo without importer in pulp2, but in case we do, it will prevent us from further remote creation, as @goosemania  mentioned, it would make sense to accumulate missing entries.